control-plane/connect-inject/controllers/pod/pod_controller_test.go

```diff
@@ -0,0 +1,1035 @@
+			pod: func() corev1.Pod {
+				pod1 := createPod("pod1", "1.2.3.4", "foo", true, true)
+				return *pod1
+			},
+			expected: true,
+		},
+		{
+			name: "Pod without injected annotation",
+			pod: func() corev1.Pod {
+				pod1 := createPod("pod1", "1.2.3.4", "foo", false, true)
+				return *pod1
+			},
+			expected: false,
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+
+			actual := hasBeenInjected(tt.pod())
+			require.Equal(t, tt.expected, actual)
+		})
+	}
+}
+
+func TestParseLocality(t *testing.T) {
+	t.Run("no labels", func(t *testing.T) {
+		n := corev1.Node{}
+		require.Nil(t, parseLocality(n))
+	})
+
+	t.Run("zone only", func(t *testing.T) {
+		n := corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					corev1.LabelTopologyZone: "us-west-1a",
+				},
+			},
+		}
+		require.Nil(t, parseLocality(n))
+	})
+
+	t.Run("everything", func(t *testing.T) {
+		n := corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					corev1.LabelTopologyRegion: "us-west-1",
+					corev1.LabelTopologyZone:   "us-west-1a",
+				},
+			},
+		}
+		require.True(t, proto.Equal(&pbcatalog.Locality{Region: "us-west-1", Zone: "us-west-1a"}, parseLocality(n)))
+	})
+}
+
+func TestWorkloadWrite(t *testing.T) {
+	t.Parallel()
+
+	ns := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{
+		Name:      metav1.NamespaceDefault,
+		Namespace: metav1.NamespaceDefault,
+	}}
+	node := corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: nodeName}}
+	localityNode := corev1.Node{ObjectMeta: metav1.ObjectMeta{
+		Name:      localityNodeName,
+		Namespace: metav1.NamespaceDefault,
+		Labels: map[string]string{
+			corev1.LabelTopologyRegion: "us-east1",
+			corev1.LabelTopologyZone:   "us-east1-b",
+		},
+	}}
+
+	type testCase struct {
+		name             string
+		pod              *corev1.Pod
+		podModifier      func(pod *corev1.Pod)
+		expectedWorkload *pbcatalog.Workload
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		if tc.podModifier != nil {
+			tc.podModifier(tc.pod)
+		}
+
+		k8sObjects := []runtime.Object{
+			&ns,
+			&node,
+			&localityNode,
+		}
+
+		fakeClient := fake.NewClientBuilder().WithRuntimeObjects(k8sObjects...).Build()
+
+		// Create test consulServer server.
+		testClient := test.TestServerWithMockConnMgrWatcher(t, func(c *testutil.TestServerConfig) {
+			c.Experiments = []string{"resource-apis"}
+		})
+		resourceClient, err := consul.NewResourceServiceClient(testClient.Watcher)
+		require.NoError(t, err)
+
+		// Create the pod controller.
+		pc := &Controller{
+			Client:                fakeClient,
+			Log:                   logrtest.New(t),
+			ConsulClientConfig:    testClient.Cfg,
+			ConsulServerConnMgr:   testClient.Watcher,
+			AllowK8sNamespacesSet: mapset.NewSetWith("*"),
+			DenyK8sNamespacesSet:  mapset.NewSetWith(),
+			ResourceClient:        resourceClient,
+		}
+
+		err = pc.writeWorkload(context.Background(), *tc.pod)
+		require.NoError(t, err)
+
```

Comment:
Does there need to be a ResourceHasPersisted check here before the read in L175?

Reply:
I _think_ it's not needed here because we are doing a read after write, and in this case, we don't care if the Consul controller has reconciled the resource yet. In the other cases, we are overwriting/deleting resources, which had the CAS behavior at the time of testing.

Reply:
I could also totally be mistaken, so please correct me. I didn't see any flakes with the test, unlike the others.

Reply:
>In the other cases, we are overwriting/deleting resources, which had the CAS behavior at the time of testing.

Ah ok, so the delete wasn't going through sometimes, unless you know consul has persisted? As opposed to in this case, the read is being done from some higher level client/cache?

I think that makes sense, I just wanted to understand the difference between this case and the others! If you ran something like a -count=100 without flakes I think it's fine!